### PR TITLE
Support Hyperlinks in SARIF explorer

### PR DESCRIPTION
**Pull request by yongyan-gh**
Current extension display SARIF result's message and rule description text in SARIF explorer. If the text contains markdown hyperlinks, it either displays the raw markdown text or the hyperlink does not work.

The change is to support display clickable hyperlink text in SARIF explorer, the hyperlink will work the same way as in error lists.

Below are screenshots show difference before/after the change.
image

### CODE REVIEW

File: src/Sarif.Viewer.VisualStudio.Core/Converters/CollectionToVisibilityConverter.cs [hunk 1]
```diff
@@ -30,4 +30,27 @@ public object ConvertBack(object value, Type targetType, object parameter, Syste
             return null;
         }
     }
+
+    public class CollectionToInvertedVisibilityConverter : IValueConverter
+    {
+        public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
```
Comment (shaopeng-gh) [inline src/Sarif.Viewer.VisualStudio.Core/Converters/CollectionToVisibilityConverter.cs:36]: >System.Globalization.CultureInfo [](http://example.com/codeflow?start=79&length=32)

add using and shorten the type?

Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Converters/CollectionToVisibilityConverter.cs]: done

File: src/Sarif.Viewer.VisualStudio.Core/ErrorList/SarifResultTableEntry.cs [hunk 2]
```diff
@@ -108,38 +103,27 @@ public SarifResultTableEntry(SarifErrorListItem error)
...
-                if (inlines.Count > 0)
-                {
-                    return inlines;
-                }
-
-                return null;
-            });
+                this.Error.MessageInlines?.Any() == true ?
+                SdkUIUtilities.GetMessageInlines(this.Error.RawMessage, this.Error.MessageInlineLink_Click) :
+                null);
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/ErrorList/SarifResultTableEntry.cs:113]: wrap using the following format:
condition
? true
: false

i think VS suggests this wrap method.

Comment (eddynaka) [reply src/Sarif.Viewer.VisualStudio.Core/ErrorList/SarifResultTableEntry.cs:113]: apply this strategy for the following two conditions below.

File: src/Sarif.Viewer.VisualStudio.Core/Models/RuleModel.cs [hunk 3]
```diff
@@ -143,5 +150,48 @@ public string DisplayName
...
+        }
+
+        public ObservableCollection<XamlDoc.Inline> DescriptionInlines
+        {
+            get
+            {
+                return this._descriptionInlines ??=
+                    string.IsNullOrWhiteSpace(this.Description) ?
+                    null :
+                    new ObservableCollection<XamlDoc.Inline>(SdkUIUtilities.GetMessageInlines(this.Description, this.InlineLink_Click));
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/Models/RuleModel.cs:169]: wrap using VS wrapping method
condition
? true
: false

File: src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs [hunk 4]
```diff
@@ -764,5 +767,69 @@ internal IEnumerable<string> GetCodeSnippets()
...
+            {
+                // The user clicked an in-line link with an integer target. Look for a Location object
+                // whose Id property matches that integer. The spec says that might be _any_ Location
+                // object under the current result. At present, we only support Location objects that
+                // occur in Result.Locations or Result.RelatedLocations. So, for example, we don't
+                // look in Result.CodeFlows or Result.Stacks.
+                LocationModel location =
+                    this.RelatedLocations.
+                    Concat(this.Locations).
+                    FirstOrDefault(l => l.Id == id);
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs:789]: this wrap is strange. pls review it.
normally the `.` is in the start.

try to use the default wrapping from VS

File: src/Sarif.Viewer.VisualStudio.UnitTests/Converters/CollectionToVisibilityConverterTests.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+using Xunit;
+
+namespace Microsoft.Sarif.Viewer.VisualStudio.Converters.UnitTests
+{
+    public class CollectionToVisibilityConverterTests
+    {
+        [Fact]
+        public void CollectionToVisibilityConverter_CollectionHasElements()
+        {
+            List<int> testList = new List<int>();
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.UnitTests/Converters/CollectionToVisibilityConverterTests.cs:23]: >List<int> [](http://example.com/codeflow?start=12&length=9)

use var

Comment (eddynaka) [reply src/Sarif.Viewer.VisualStudio.UnitTests/Converters/CollectionToVisibilityConverterTests.cs:23]: update all the other places where in this file.

File: src/Sarif.Viewer.VisualStudio.Core/ErrorList/SarifResultTableEntry.cs (hunk 2 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/ErrorList/SarifResultTableEntry.cs]: updated all conditions format

File: src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs (hunk 4 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs]: used VS wrapping format

File: src/Sarif.Viewer.VisualStudio.UnitTests/Converters/CollectionToVisibilityConverterTests.cs (hunk 5 above)
Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.UnitTests/Converters/CollectionToVisibilityConverterTests.cs]: updated all

**Review by eddynaka — APPROVED** [review]
:shipit:

File: src/Sarif.Viewer.VisualStudio.UnitTests/ErrorList/SarifResultEntryTests.cs [hunk 6]
```diff
@@ -58,5 +65,55 @@ public void SarifSnapshot_TryGetValue_NegativeLineNumber()
             tableEntry.TryGetValue("line", out value).Should().Be(true);
             value.Should().Be(lineNumber - 1);
         }
+
+        [Fact]
+        public void SarifResultEntry_TryGetValue_TextWithMarkdownHyperlinks()
+        {
+            string messageText = "Sample [text](0) with hyperlink to [github](https://github.com).";
```
Comment (michaelcfanning) [inline src/Sarif.Viewer.VisualStudio.UnitTests/ErrorList/SarifResultEntryTests.cs:72]: >github [](http://example.com/codeflow?start=86&length=6)

We typically use example.com for URLs in tests.

**Pull request merged**